store/store.go

```diff
@@ -499,6 +506,67 @@ func (s *Store[H]) get(ctx context.Context, hash header.Hash) ([]byte, error) {
 	return data, nil
 }
 
+// notifyAndAdvance will notify waiters in heightSub and advance contiguousHead
+// based on already written headers.
+func (s *Store[H]) notifyAndAdvance(ctx context.Context, headers ...H) {
+	// always inform heightSub about new headers seen
+	for _, h := range headers {
+		s.heightSub.Notify(h.Height())
+	}
+
+	currHead := s.contiguousHead.Load()
+	if currHead != nil {
+		s.advanceContiguousHead(ctx, (*currHead).Height())
+	}
+}
+
+// advanceContiguousHead return a new highest contiguous height
+// or returns the given height if not found.
+func (s *Store[H]) advanceContiguousHead(ctx context.Context, currHeight uint64) uint64 {
+	// TODO(cristaloleg): benchmark this timeout or make it dynamic.
+	advCtx, advCancel := context.WithTimeout(ctx, 10*time.Second)
+	defer advCancel()
+
+	prevHeight := currHeight
+	var newHead H
+	for {
+		h, err := s.getByHeight(advCtx, currHeight+1)
+		if err != nil {
+			break
+		}
+		newHead = h
+		currHeight++
+	}
+
+	if currHeight > prevHeight {
+		s.updateContiguousHead(ctx, newHead)
+	}
+	return currHeight
+}
+
+func (s *Store[H]) updateContiguousHead(ctx context.Context, newHead H) {
+	s.contiguousHead.Store(&newHead)
+	s.heightSub.SetHeight(newHead.Height())
+	log.Infow("new head", "height", newHead.Height(), "hash", newHead.Hash())
+	s.metrics.newHead(newHead.Height())
+
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
```

Comment:
If we end up inheriting contexts in the end, there is no need in doing children in this case.

Reply:
actually nvm